File: 6_kyu/Equal_Sides_Of_An_Array.go
package kata

// Sum of all integers in arr.
// LIMITATION: sums must in int64 range.
func sum(arr []int) int64 {
  var accu int64 = 0
  for _, it := range arr {
    accu += int64(it)
  }
  return accu
}

// Find index i so that arr[:i] = arr[i+1:].
// Item i is not part of the sums.
// Returns -1 if no such index exists.
func FindEvenIndex(arr []int) int {
  if len(arr) == 0 {
    return 0
  }
  // Integer arithmetix is fast. So no caching of the partial sums.
  s := sum(arr)
  var left int64 = 0
  var right int64 = s
  for index, it := range arr {
    it64 := int64(it)
    right -= it64
    if left == right {
      return index
    }
    left += it64
  }
  return -1
}

File: 6_kyu/Equal_Sides_Of_An_Array.kt
object EqualSidesOfAnArray {
    /**
     * Finds index with same item sum right and left of the item. Limitation: Sum and all parital
     * sums must be in the Int range.
     *
     * @param arr Search in this array
     * @returns index with sum(arr[..index-1]) == sum(arr[index+1..]) or -1.
     */
    fun findEvenIndex(arr: IntArray): Int {
        val total = arr.sum()
        var accu = 0
        return arr.indexOfFirst { it ->
            val cmp = 2 * accu == total - it
            accu += it
            cmp
        }
    }
}
